module/zfs/dmu_object.c

```diff
@@ -236,34 +236,39 @@ int
 dmu_object_next(objset_t *os, uint64_t *objectp, boolean_t hole, uint64_t txg)
 {
 	uint64_t offset;
-	dmu_object_info_t doi;
+	uint64_t start_obj;
 	struct dsl_dataset *ds = os->os_dsl_dataset;
-	int dnodesize;
 	int error;
 
-	/*
-	 * Avoid expensive dnode hold if this dataset doesn't use large dnodes.
-	 */
-	if (ds && ds->ds_feature_inuse[SPA_FEATURE_LARGE_DNODE]) {
-		error = dmu_object_info(os, *objectp, &doi);
+	if (*objectp == 0) {
+		start_obj = 1;
+	} else if (ds && ds->ds_feature_inuse[SPA_FEATURE_LARGE_DNODE]) {
 		/*
-		 * dmu_object_next() consumers expect ESRCH if the object is not
-		 * allocated.
+		 * For large_dnode datasets, scan from the beginning of the
+		 * dnode block to find the starting offset. This is needed
+		 * because objectp could be part of a large dnode so we can't
+		 * assume it's a hole even if dmu_object_info() returns ENOENT.
 		 */
-		if (error == ENOENT)
-			return (SET_ERROR(ESRCH));
-		else if (error && !(error == EINVAL && *objectp == 0))
-			return (SET_ERROR(error));
-		else
-			dnodesize = doi.doi_dnodesize;
+		int epb = DNODE_BLOCK_SIZE >> DNODE_SHIFT;
+		int skip;
+		uint64_t i;
+
+		for (i = *objectp & ~(epb - 1); i <= *objectp; i += skip) {
+			dmu_object_info_t doi;
+
+			error = dmu_object_info(os, i, &doi);
+			if (error)
+				skip = 1;
+			else
+				skip = doi.doi_dnodesize >> DNODE_SHIFT;
+		}
+
+		start_obj = i;
```

Comment:
Originally I was going to suggest refactoring this in to a helper function since it's a generically useful functionality for working with large dnodes.  But upon further reflection I don't think it's necessary until we have another consumer.